### ISSUE TITLE
[clang] No lazy evaluation of boolean alternative in a constant-evaluated expression

### ISSUE THREAD

**Issue by dawmd** [llvm/llvm-project#93750]
Clang rejects this code:

```cpp
#include <concepts>

template <typename T, bool AlreadyEvaluated = false>
struct is_addable {
    static constexpr bool value = AlreadyEvaluated || requires (T t) {
        requires is_addable<decltype(t + t), std::same_as<decltype(t + t), T>>::value;
    };
};

int main() {
    static_assert(is_addable<int>::value, "");
}
```

Flags used to compile the code: `-std=c++23`
Clang version: 19.0.0 llvm/llvm-project@5bfe4b93e15ad38f211c5dec64be0eeaa4c8e914

**Comment by zyn0217** [comment]
I don't think we should accept the code...? Note GCC gives us a clearer explanation in that
```
<source>:6:81: error: satisfaction of atomic constraint 'is_addable<decltype(t + t), same_as<decltype(t + t), T> >::value [with T = T]' depends on itself
```
https://gcc.godbolt.org/z/GYheW5Mhe

@shafik @erichkeane Thoughts?

**Comment by cor3ntin** [comment]
Yes, this is ill formed. In some contexts (such as when checking a constraint), short circuiting occurs, such that expressions that don't need to be evaluated are not instantiated.

However, here `bool value = AlreadyEvaluated || requires { /* */}` is not a constraint but a regular boolean expression - so both branches get instantiated recursively.
You can rewrite the code slightly to get the short circuiting behavior you are after https://gcc.godbolt.org/z/5v77az46d
